ipn/ipnlocal/local.go

```diff
@@ -3522,18 +3523,99 @@ func (b *LocalBackend) setTCPPortsInterceptedFromNetmapAndPrefsLocked() {
 				}
 			}
 			if b.serveConfig.Valid() {
+				servePorts := make([]uint16, 0, 3)
 				b.serveConfig.TCP().Range(func(port uint16, _ ipn.TCPPortHandlerView) bool {
 					if port > 0 {
-						handlePorts = append(handlePorts, uint16(port))
+						servePorts = append(servePorts, uint16(port))
 					}
 					return true
 				})
+				handlePorts = append(handlePorts, servePorts...)
+				// don't listen on netmap addresses if we're in userspace mode
+				if !wgengine.IsNetstack(b.e) {
+					b.updateServeTCPPortNetMapAddrListenersLocked(servePorts)
+				}
 			}
 		}
 	}
 	b.setTCPPortsIntercepted(handlePorts)
 }
 
+// updateServeTCPPortNetMapAddrListenersLocked starts a net.Listen for configured
+// Serve ports on all the node's addresses.
+// Existing Listeners are closed if port no longer in incoming ports list.
+//
+// b.mu must be held.
+func (b *LocalBackend) updateServeTCPPortNetMapAddrListenersLocked(ports []uint16) {
+	nm := b.netMap
+	if nm == nil {
+		b.logf("netMap is nil")
+		return
+	}
+	if nm.SelfNode == nil {
+		b.logf("netMap SelfNode is nil")
+		return
+	}
+
+	if b.serveListeners == nil {
+		b.serveListeners = make(map[netip.AddrPort]net.Listener)
+	}
+
+	for ap, ln := range b.serveListeners {
```

Comment:
this loop needs to move to the very top, before the `nm == nil` early return, no?